content/courses/program-security/reinitialization-attacks.md

```diff
@@ -114,260 +103,268 @@ pub mod initialization_secure {
-via a CPI to the system program and sets the account discriminator.
-
-The `init` constraint must be used in combination with the `payer` and `space`
-constraints. The `payer` specifies the account paying for the initialization of
-the new account. The `space` specifies the amount of space the new account
-requires, which determines the amount of lamports that must be allocated to the
-account. The first 8 bytes of data is set as a discriminator that Anchor
-automatically adds to identify the account type.
-
-Most importantly for this lesson, the `init` constraint ensures that this
-instruction can only be called once per account, so you can set the initial
-state of the account in the instruction logic and not have to worry about an
-attacker trying to reinitialize the account.
+Anchor’s `init` constraint, used with the `#[account(...)]` attribute,
+initializes an account, sets the account discriminator, and ensures that the
+instruction can only be called once per account. The `  ` constraint must be
+used with `payer` and `space` constraints to specify the account paying for
+initialization and the amount of space required.
 
 ```rust
 use anchor_lang::prelude::*;
 
 declare_id!("Fg6PaFpoGXkYsidMpWTK6W2BeZ7FEfcYkg476zPFsLnS");
 
+const DISCRIMINATOR_SIZE: usize = 8;
+
 #[program]
 pub mod initialization_recommended {
     use super::*;
 
-    pub fn initialize(_ctx: Context<Initialize>) -> Result<()> {
+    pub fn initialize(ctx: Context<Initialize>) -> Result<()> {
         msg!("GM");
+        ctx.accounts.user.authority = ctx.accounts.authority.key();
         Ok(())
     }
 }
 
 #[derive(Accounts)]
 pub struct Initialize<'info> {
-    #[account(init, payer = authority, space = 8+32)]
-    user: Account<'info, User>,
+    #[account(
+        init,
+        payer = authority,
+        space = DISCRIMINATOR_SIZE + User::INIT_SPACE
+    )]
+    pub user: Account<'info, User>,
     #[account(mut)]
-    authority: Signer<'info>,
-    system_program: Program<'info, System>,
+    pub authority: Signer<'info>,
+    pub system_program: Program<'info, System>,
 }
 
 #[account]
+#[derive(InitSpace)]
 pub struct User {
-    authority: Pubkey,
+    pub authority: Pubkey,
 }
 ```
 
-#### Anchor’s `init_if_needed` constraint
-
-It’s worth noting that Anchor has an `init_if_needed` constraint. This
-constraint should be used very cautiously. In fact, it is blocked behind a
-feature flag so that you are forced to be intentional about using it.
+#### Anchor’s init_if_needed Constraint
 
-The `init_if_needed` constraint does the same thing as the `init` constraint,
-only if the account has already been initialized the instruction will still run.
+Anchor’s `init_if_needed` constraint, guarded by a feature flag, should be used
+with caution. It initializes an account only if it hasn’t been initialized yet.
+If the account is already initialized, the instruction will still run, so
+it's \***\*\*\*\***extremely\***\*\*\*\*** important to include checks to
```

Comment:
Since this is security related, you might want to make [callout](https://github.com/solana-foundation/developer-content/blob/main/CONTRIBUTING.md#callout) for this.

Reply:
I added a callout for the whole text as a **caution** to emphasize the security.